### DISCUSSION TITLE
Automatic awaiting of tasks

### DISCUSSION THREAD

**Discussion by VerticalVeith** [dotnet/csharplang#5821] [Language Ideas] [answered]
What potential problems could arise from automatically awaiting asynchronous tasks?

# Motivation
Asynchronous Functions/Methods can lead to unexpected behavior when they are not awaited. 
Consider the following:
there are two asynchronous methods. 

`GetMagicString` is used to generate some magical string asynchronously.
`LogState` logs the machine state using some kind of logger and may take a couple of minutes to complete.

```csharp
async Task<string> GetMagicString()
{
    var text = await GenerateSomeTextAsync();
    text.ApplyMagic();
    return text;
}

async Task LogState()
{
    await _horribleLogger.LogStateAsync();
}
```

Now, if there is a block of code somewhere like this:

```csharp
{
  var magic = GetMagicString();
  LogState();
  Console.WriteLine(magic);
}
```
`LogState` could complete without any issues as long as it does not run into a deadlock or throw an exception.
`GetMagicString` however, will log something along the lines of ``System.Threading.Tasks.Task`1[System.String]`` instead of the expected string.

Although analyzers do catch the first issue and some could catch the second, that is not a guarantee.

# The proposal
To avoid these kinds of problems, C# could introduce an _optional_ feature, which would automatically await a task when trying to get its value, or it goes out of scope.

Essentially, it could rewrite the last part to something like this:
```csharp
{
  var magic = GetMagicString();
  var task = LogState();
  
  Console.WriteLine((await magic));
  
  await task;
}
```

This is nothing more than an idea that has been bugging me for the last couple of day. Obviously, I do not wish for this feature to be added in its current form. It would interest me what kind of difficulties this could cause, and if there even is a place for this feature. I tried to research if any other language has implemented anything like this, but I failed to find anything regarding this matter.

**Comment by HaloFour** [comment, marked as answer]
See: https://github.com/dotnet/csharplang/discussions/5469
